src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,706 @@
+        }
+
+        if (this->currentHeader->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD)
+        {
+            // This record has been overwritten.
+            // Write to the output what we managed to read as is, skipping filtering.
+            return ReadRecordSuccess;
+        }
+
+        if (!(this->currentHeader->xlp_info & XLP_FIRST_IS_CONTRECORD))
+        {
+            THROW_FMT(FormatError, "%s - should be XLP_FIRST_IS_CONTRECORD", strZ(pgLsnToStr(this->recPtr)));
+        }
+
+        memcpy(((char *) this->record) + this->gotLen, this->page + this->pageOffset, SizeOfXLogRecord - this->gotLen);
+        this->totLen -= this->gotLen;
+        this->headerOffset = SizeOfXLogRecord - this->gotLen;
+    }
+    else
+    {
+        this->headerOffset = SizeOfXLogRecord;
+    }
+    this->gotLen = SizeOfXLogRecord;
+
+    this->walInterface->validXLogRecordHeader(this->record, this->heapPageSize);
+    // Read rest of the record on this page
+    size_t to_read = Min(this->record->xl_tot_len - SizeOfXLogRecord, this->walPageSize - this->pageOffset - SizeOfXLogRecord);
+    memcpy(XLogRecGetData(this->record), this->page + this->pageOffset + this->headerOffset, to_read);
+    this->gotLen += to_read;
+
+    // Move pointer to the next record on the page
+    this->pageOffset += MAXALIGN(this->totLen);
+
+    // Rest of the record data is on the next page
+    while (this->gotLen != this->record->xl_tot_len)
+    {
+stepReadBody:
+        if (!readPage(this, input, stepReadBody))
+        {
+            return ReadRecordNeedBuffer;
+        }
+
+        if (this->currentHeader->xlp_info & XLP_FIRST_IS_OVERWRITE_CONTRECORD)
+        {
+            // This record has been overwritten.
+            // Write to the output what we managed to read as is, skipping filtering.
+            return ReadRecordSuccess;
+        }
+
+        if (!(this->currentHeader->xlp_info & XLP_FIRST_IS_CONTRECORD))
+        {
+            THROW_FMT(FormatError, "%s - should be XLP_FIRST_IS_CONTRECORD", strZ(pgLsnToStr(this->recPtr)));
+        }
+
+        if (this->currentHeader->xlp_rem_len == 0 ||
+            this->totLen != (this->currentHeader->xlp_rem_len + this->gotLen))
+        {
+            THROW_FMT(FormatError, "%s - invalid contrecord length: expect: %zu, get %u", strZ(pgLsnToStr(this->recPtr)),
+                      this->record->xl_tot_len - this->gotLen, this->currentHeader->xlp_rem_len);
+        }
+
+        size_t to_write = Min(this->currentHeader->xlp_rem_len, this->walPageSize - this->pageOffset);
+        memcpy(((char *) this->record) + this->gotLen, this->page + this->pageOffset, to_write);
+        this->pageOffset += MAXALIGN(to_write);
+        this->gotLen += to_write;
+    }
+    this->walInterface->validXLogRecord(this->record, this->heapPageSize);
+
+    if (this->record->xl_rmid == RM_XLOG_ID && this->record->xl_info == XLOG_SWITCH)
+    {
+        this->isSwitchWal = true;
+    }
+    this->recordNum++;
```

Comment:
What is the point of this variable if it is only incremented before this `ReadRecordSuccess` return?

Reply:
When debugging, it may be useful to know which record the error occurred on.